aptos-move/block-executor/src/cross_block_caches.rs

```diff
@@ -0,0 +1,413 @@
+    /// Returns true if the entry is valid.
+
+    pub fn is_valid(&self) -> bool {
+        self.valid.load(Ordering::Acquire)
+    }
+
+    /// Returns the state value metadata if the entry is valid, and [None] otherwise.
+    fn state_value_metadata(&self) -> Option<StateValueMetadata> {
+        self.is_valid()
+            .then(|| self.verified_entry.state_value_metadata().clone())
+    }
+
+    /// Returns the module bytes if the entry is valid, and [None] otherwise.
+    fn bytes(&self) -> Option<Bytes> {
+        self.is_valid().then(|| self.verified_entry.bytes().clone())
+    }
+
+    /// Returns the module size in bytes if the entry is valid, and [None] otherwise.
+    fn size_in_bytes(&self) -> Option<usize> {
+        self.is_valid().then(|| self.verified_entry.size_in_bytes())
+    }
+
+    /// Returns the module metadata if the entry is valid, and [None] otherwise.
+    fn module_metadata(&self) -> Option<Vec<Metadata>> {
+        self.is_valid()
+            .then(|| self.verified_entry.metadata().to_vec())
+    }
+
+    /// Returns the deserialized module if the entry is valid, and [None] otherwise.
+    fn deserialized_module(&self) -> Option<Arc<CompiledModule>> {
+        self.is_valid()
+            .then(|| self.verified_entry.compiled_module().clone())
+    }
+
+    /// Returns the verified module if the entry is valid, and [None] otherwise. Panics if the
+    /// entry is not verified.
+    fn verified_module(&self) -> VMResult<Option<Arc<Module>>> {
+        if self.is_valid() {
+            self.verified_entry
+                .verified_module()
+                .map(|v| Some(v.clone()))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+/// Represents an immutable cross-block cache. The size of the cache is fixed (entries cannot be
+/// added or removed) within a single block, so it is only mutated at block boundaries. At the
+/// same time, entries in this cache can be marked as "invalid" so that block executor can decide
+/// on whether to read the module from cache or from the storage.
+pub struct CrossBlockModuleCache;
+
+impl CrossBlockModuleCache {
+    /// Flushes the module cache. Should only be called at the start of the block.
+    pub fn flush_at_block_start() {
+        let mut cache = CROSS_BLOCK_MODULE_CACHE.acquire();
+        cache.clear();
+    }
+
+    /// Adds new verified entries from block-level cache to the cross-block cache. Flushes the
+    /// cache if its size is too large. Should only be called at block end.
+    pub(crate) fn populate_from_sync_code_cache_at_block_end(
+        code_cache: &SyncCodeCache<
+            ModuleId,
+            Arc<MaybeCommitted<ModuleCacheEntry>>,
+            [u8; 32],
+            CachedScript,
+            VMError,
+        >,
+    ) {
+        let mut cache = CROSS_BLOCK_MODULE_CACHE.acquire();
+        if cache.len() > MAX_CROSS_BLOCK_MODULE_CACHE_SIZE {
+            cache.clear();
+        }
+
+        code_cache.module_cache().filter_into(
```

Comment:
we talked about avoiding the cloning here, also this seems the only place we use the "filter_into", probably not worth the generics.

Reply:
Changed code so that one can consume module cache (unsync) or take modules (sync) to pass here. Removed `filter_into` and duplication between populating global cache with sync/unsync calls.